pkg/tracer/offsetguess.go

```diff
@@ -0,0 +1,355 @@
+	for i := 0; i < 16; i++ {
+		buf[i] = *(*byte)(unsafe.Pointer((uintptr(unsafe.Pointer(&ipv6Addr[0])) + uintptr(i))))
+	}
+	return net.IP(buf)
+}
+
+func htons(a uint16) uint16 {
+	var arr [2]byte
+	binary.BigEndian.PutUint16(arr[:], a)
+	return nativeEndian.Uint16(arr[:])
+}
+
+// tryCurrentOffset creates a IPv4 or IPv6 connection so the corresponding
+// tcp_v{4,6}_connect kprobes get triggered and save the value at the current
+// offset in the eBPF map
+func tryCurrentOffset(module *elf.Module, mp *elf.Map, status *tcpTracerStatus, expected *fieldValues) error {
+	// for ipv6, we don't need the source port because we already guessed
+	// it doing ipv4 connections so we use a random destination address and
+	// try to connect to it
+	expected.daddrIPv6[0] = rand.Uint32()
+	expected.daddrIPv6[1] = rand.Uint32()
+	expected.daddrIPv6[2] = rand.Uint32()
+	expected.daddrIPv6[3] = rand.Uint32()
+
+	ip := ipv6FromUint32Arr(expected.daddrIPv6)
+
+	bindAddress := fmt.Sprintf("%s:%d", listenIP, expected.dport)
+	if status.what != guessDaddrIPv6 {
+		conn, err := net.Dial("tcp4", bindAddress)
+		if err != nil {
+			return fmt.Errorf("error dialing %q: %v\n", bindAddress, err)
+		}
+
+		// get the source port assigned by the kernel
+		sport, err := strconv.Atoi(strings.Split(conn.LocalAddr().String(), ":")[1])
+		if err != nil {
+			return fmt.Errorf("error converting source port: %v", err)
+		}
+
+		expected.sport = uint16(sport)
+
+		// set SO_LINGER to 0 so the connection state after closing is
+		// CLOSE instead of TIME_WAIT. In this way, they will disappear
+		// from the conntrack table after around 10 seconds instead of 2
+		// minutes
+		if tcpConn, ok := conn.(*net.TCPConn); ok {
+			tcpConn.SetLinger(0)
+		} else {
+			return fmt.Errorf("not a tcp connection unexpectedly")
+		}
+
+		conn.Close()
+	} else {
+		conn, err := net.Dial("tcp6", fmt.Sprintf("[%s]:9092", ip))
+		// Since we connect to a random IP, this will most likely fail.
+		// In the unlikely case where it connects successfully, we close
+		// the connection to avoid a leak.
+		if err == nil {
+			conn.Close()
+		}
+	}
+
+	return nil
+}
+
+// checkAndUpdateCurrentOffset checks the value for the current offset stored
+// in the eBPF map against the expected value, incrementing the offset if it
+// doesn't match, or going to the next field to guess if it does
+func checkAndUpdateCurrentOffset(module *elf.Module, mp *elf.Map, status *tcpTracerStatus, expected *fieldValues) error {
+	// get the updated map value so we can check if the current offset is
+	// the right one
+	if err := module.LookupElement(mp, unsafe.Pointer(&zero), unsafe.Pointer(status)); err != nil {
+		return fmt.Errorf("error reading tcptracer_status: %v", err)
+	}
+
+	if status.status != checked {
+		return fmt.Errorf("invalid guessing state")
+	}
+
+	switch status.what {
+	case guessSaddr:
+		if status.saddr == C.__u32(expected.saddr) {
+			status.what = guessDaddr
+			status.status = checking
```

Comment:
@brb I'm not sure to understand: who would you do that? With 7 small functions or with one parameterized function somehow?